python/ml4ir/base/pipeline.py

```diff
@@ -265,9 +269,209 @@ def get_relevance_model(self, feature_layer_keys_to_fns={}) -> RelevanceModel:
+            Used for model selection and hyperparameter optimization
+
+        Notes
+        -----
+        Also populates a experiment tracking dictionary containing
+        the metadata, model architecture and metrics generated by the model
+        """
+        if self.args.kfold <= 1:
+            # Run ml4ir without kfold cross validation
+            return self.run_pipeline()
+
+        if self.args.include_testset_in_kfold:
+            if self.args.kfold < 3:
+                raise Exception("Number of folds must be > 2")
+        else:
+            if self.args.kfold < 2:
+                raise Exception("Number of folds must be > 1")
+
+        args = copy.deepcopy(self.args)
+
+        # reading, parsing the dataset (train, validation, test)
+        self.logger.info("Reading datasets ...")
+        relevance_dataset = self.get_relevance_dataset()
+        self.logger.info("Relevance Dataset created")
+
+        all_data = self.merge_datasets(relevance_dataset, args.include_testset_in_kfold)
+
+        num_folds = self.args.kfold
+        base_logs_dir = str(self.args.logs_dir)
+        base_models_dir = str(self.args.models_dir)
+        base_run_id = self.run_id
+        self.logger.info("K-fold Cross Validation mode starting with k={}".format(self.args.kfold))
+        self.logger.info("Include testset in the folds={}".format(str(self.args.include_testset_in_kfold)))
+
+        # when creating folds, the validation set is assigned fold i, test fold i+1 and training get the rest of folds
+        for i in range(num_folds):
+            self.logger.info("fold={}".format(i))
+            logs_dir = pathlib.Path(base_logs_dir) / self.args.run_id / "fold_{}".format(i)
+            models_dir = pathlib.Path(base_models_dir) / self.args.run_id / "fold_{}".format(i)
+            args.logs_dir = pathlib.Path(logs_dir).as_posix()
+            args.models_dir = pathlib.Path(models_dir).as_posix()
+
+            training, validation, test = self.create_folds(num_folds, i, args.include_testset_in_kfold, all_data)
+
+            # batchify training, validation and test sets.
+            validation = validation.batch(args.batch_size, drop_remainder=False)
+            training = training.batch(args.batch_size, drop_remainder=False)
+            # We apply prefetch as it improved train/test/validation throughput by 30% in some real model training.
+            relevance_dataset.training = training.prefetch(tf.data.experimental.AUTOTUNE)
+            relevance_dataset.validation = validation.prefetch(tf.data.experimental.AUTOTUNE)
+            if args.include_testset_in_kfold:
+                test = test.batch(args.batch_size, drop_remainder=False)
+                relevance_dataset.test = test.prefetch(tf.data.experimental.AUTOTUNE)
+
+            # run the pipeline over the created folds.
+            pipeline = self.create_pipeline_for_kfold(args)
+            pipeline.run_pipeline(relevance_dataset=relevance_dataset)
```

Comment:
@mohazahran I don't think you've still addressed the issue of creating multiple models. It will definitely blow up memory I have tried this.

Reply:
I thought the memory issues were related to creating multiple RelevanceDatasets.
will look into that